'PySys_SetPath' is deprecated warning from  SetupLocalPythonEnvrionment

# Bug reports:

### Expected behavior:

Code compiles with no warnings

### Current behavior:

```
[879/999] Building CXX object tests/Unit/Framework/CMakeFiles/Framework.dir/SetupLocalPythonEnvironment.cpp.o
/home/lek/Research/Codes/spectre/kidder/feature/random_h_refinement/tests/Unit/Framework/SetupLocalPythonEnvironment.cpp:63:5: warning: 'PySys_SetPath' is deprecated [-Wdeprecated-declarations]
    PySys_SetPath(std::wstring_convert<std::codecvt_utf8<wchar_t>>()
    ^
/usr/include/python3.11/sysmodule.h:15:1: note: 'PySys_SetPath' has been explicitly marked deprecated here
Py_DEPRECATED(3.11) PyAPI_FUNC(void) PySys_SetPath(const wchar_t *);
^
/usr/include/python3.11/pyport.h:336:54: note: expanded from macro 'Py_DEPRECATED'
#define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
                                                     ^
1 warning generated.
```

### Environment:

Python 3.11.4 on my Fedora 38 desktop

### Detailed discussion:

Just a warning for now, so nothing critical

--- a/tests/Unit/Framework/CheckWithRandomValues.hpp
+++ b/tests/Unit/Framework/CheckWithRandomValues.hpp
@@ -163,20 +163,18 @@ void check_with_random_values_impl(
     using Tag = tmpl::type_from<decltype(tag)>;
     const auto result =
         tuples::get<Tag>((klass.*f)(std::get<ArgumentIs>(args)...));
-    INFO("function: " << function_names[count]);
-    try {
-      CHECK_ITERABLE_CUSTOM_APPROX(
-          result,
-          (pypp::call<std::decay_t<decltype(result)>>(
-              module_name, function_names[count], std::get<ArgumentIs>(args)...,
-              forward_to_pypp<std::decay_t<decltype(result)>>(
-                  std::get<MemberArgsIs>(member_args), used_for_size)...)),
-          Approx::custom().epsilon(epsilon).scale(1.0));
-      count++;
-    } catch (const std::exception& e) {
-      INFO("Python call failed: " << e.what());
-      CHECK(false);
-    }
+    const auto& function_name = function_names[count];
+    CAPTURE(module_name);
+    CAPTURE(function_name);
+    CAPTURE(args);
+    CAPTURE(member_args);
+    const auto python_result = pypp::call<std::decay_t<decltype(result)>>(
+        module_name, function_name, std::get<ArgumentIs>(args)...,
+        forward_to_pypp<std::decay_t<decltype(result)>>(
+            std::get<MemberArgsIs>(member_args), used_for_size)...);
+    CHECK_ITERABLE_CUSTOM_APPROX(result, python_result,
+                                 Approx::custom().epsilon(epsilon).scale(1.0));
+    ++count;
   });
 }
 
@@ -209,19 +207,16 @@ void check_with_random_values_impl(
       return (klass.*f)(std::get<ArgumentIs>(args)...);
     }
   }();
-  INFO("function: " << function_name);
-  try {
-    CHECK_ITERABLE_CUSTOM_APPROX(
-        result,
-        pypp::call<ResultType>(
-            module_name, function_name, std::get<ArgumentIs>(args)...,
-            forward_to_pypp<ResultType>(std::get<MemberArgsIs>(member_args),
-                                        used_for_size)...),
-        Approx::custom().epsilon(epsilon).scale(1.0));
-    } catch (const std::exception& e) {
-      INFO("Python call failed: " << e.what());
-      CHECK(false);
-    }
+  CAPTURE(module_name);
+  CAPTURE(function_name);
+  CAPTURE(args);
+  CAPTURE(member_args);
+  const auto python_result = pypp::call<ResultType>(
+      module_name, function_name, std::get<ArgumentIs>(args)...,
+      forward_to_pypp<ResultType>(std::get<MemberArgsIs>(member_args),
+                                  used_for_size)...);
+  CHECK_ITERABLE_CUSTOM_APPROX(result, python_result,
+                               Approx::custom().epsilon(epsilon).scale(1.0));
 }
 
 template <class F, class T, class Klass, class... ReturnTypes,
@@ -265,20 +260,18 @@ void check_with_random_values_impl(
     (void)member_args;    // avoid compiler warning
     (void)used_for_size;  // avoid compiler warning
     constexpr size_t iter = decltype(result_i)::value;
-    INFO("function: " << function_names[iter]);
-    try {
-      CHECK_ITERABLE_CUSTOM_APPROX(
-          std::get<iter>(results),
-          (pypp::call<std::tuple_element_t<iter, std::tuple<ReturnTypes...>>>(
-              module_name, function_names[iter], std::get<ArgumentIs>(args)...,
-              forward_to_pypp<
-                  std::tuple_element_t<iter, std::tuple<ReturnTypes...>>>(
-                  std::get<MemberArgsIs>(member_args), used_for_size)...)),
-          Approx::custom().epsilon(epsilon).scale(1.0));
-    } catch (const std::exception& e) {
-      INFO("Python call failed: " << e.what());
-      CHECK(false);
-    }
+    const auto& function_name = function_names[iter];
+    CAPTURE(module_name);
+    CAPTURE(function_name);
+    CAPTURE(args);
+    CAPTURE(member_args);
+    const auto python_result = pypp::call<
+        std::tuple_element_t<iter, std::tuple<ReturnTypes...>>>(
+        module_name, function_name, std::get<ArgumentIs>(args)...,
+        forward_to_pypp<std::tuple_element_t<iter, std::tuple<ReturnTypes...>>>(
+            std::get<MemberArgsIs>(member_args), used_for_size)...);
+    CHECK_ITERABLE_CUSTOM_APPROX(std::get<iter>(results), python_result,
+                                 Approx::custom().epsilon(epsilon).scale(1.0));
   };
   EXPAND_PACK_LEFT_TO_RIGHT(helper(std::integral_constant<size_t, ResultIs>{}));
 }

--- a/tests/Unit/Framework/SetupLocalPythonEnvironment.cpp
+++ b/tests/Unit/Framework/SetupLocalPythonEnvironment.cpp
@@ -28,11 +28,17 @@ SetupLocalPythonEnvironment::SetupLocalPythonEnvironment(
   // finished running, since there could be multiple tests run in a single
   // executable launch. This is done in TestMain(Charm).cpp.
   if (not initialized) {
+    PyStatus status;
+    PyConfig config;
+    PyConfig_InitPythonConfig(&config);
+    // Populate the Python config with the standard values
+    status = PyConfig_Read(&config);
     // Don't produce the __pycache__ dir (python 3.2 and newer) or the .pyc
     // files (python 2.7) in the tests directory to avoid cluttering the source
     // tree. The overhead of not having the compile files is <= 0.01s
-    Py_DontWriteBytecodeFlag = 1;
-    Py_Initialize();
+    config.write_bytecode = 0;
+    status = Py_InitializeFromConfig(&config);
+    PyConfig_Clear(&config);
 
     // clang-tidy: Do not use const-cast
     PyObject* pyob_old_paths =

--- a/tests/Unit/Framework/SetupLocalPythonEnvironment.cpp
+++ b/tests/Unit/Framework/SetupLocalPythonEnvironment.cpp
@@ -2,8 +2,6 @@
 // See LICENSE.txt for details.
 
 #include <boost/preprocessor.hpp>
-#include <codecvt>  // IWYU pragma: keep
-#include <locale>   // IWYU pragma: keep
 #include <string>
 #include <vector>
 
@@ -18,12 +16,27 @@
 #include "Framework/PyppFundamentals.hpp"
 #include "Framework/SetupLocalPythonEnvironment.hpp"
 #include "Informer/InfoFromBuild.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
 #include "Utilities/ErrorHandling/FloatingPointExceptions.hpp"
 #include "Utilities/FileSystem.hpp"
 
+namespace {
+void prepend_python_path(const std::string& new_path) {
+  if (not file_system::check_if_dir_exists(new_path)) {
+    ERROR_NO_TRACE("Trying to add path '"
+                   << new_path
+                   << "' to the python environment during setup "
+                      "but this directory does not exist. Maybe "
+                      "you have a typo in your path?");
+  }
+  PyObject* sys_path = PySys_GetObject("path");
+  PyList_Insert(sys_path, 0, PyUnicode_FromString(new_path.c_str()));
+}
+}  // namespace
+
 namespace pypp {
 SetupLocalPythonEnvironment::SetupLocalPythonEnvironment(
-    const std::string &cur_dir_relative_to_unit_test_path) {
+    const std::string& cur_dir_relative_to_unit_test_path) {
   // We have to clean up the Python environment only after all tests have
   // finished running, since there could be multiple tests run in a single
   // executable launch. This is done in TestMain(Charm).cpp.
@@ -40,38 +53,8 @@ SetupLocalPythonEnvironment::SetupLocalPythonEnvironment(
     status = Py_InitializeFromConfig(&config);
     PyConfig_Clear(&config);
 
-    // clang-tidy: Do not use const-cast
-    PyObject* pyob_old_paths =
-        PySys_GetObject(const_cast<char*>("path"));  // NOLINT
-    const auto old_paths =
-        pypp::from_py_object<std::vector<std::string>>(pyob_old_paths);
-    std::string new_path =
-        unit_test_src_path() + cur_dir_relative_to_unit_test_path;
-    if (not file_system::check_if_dir_exists(new_path)) {
-      ERROR_NO_TRACE("Trying to add path '"
-                     << new_path
-                     << "' to the python environment during setup "
-                        "but this directory does not exist. Maybe "
-                        "you have a typo in your path?");
-    }
-
     // Add directory for installed packages (see CMakeLists.txt for details)
-    new_path += ":";
-    new_path += PYTHON_SITELIB;
-
-    for (const auto& p : old_paths) {
-      new_path += ":";
-      new_path += p;
-    }
-
-#if PY_MAJOR_VERSION == 3
-    PySys_SetPath(std::wstring_convert<std::codecvt_utf8<wchar_t>>()
-                      .from_bytes(new_path)
-                      .c_str());
-#else
-    // clang-tidy: Do not use const-cast
-    PySys_SetPath(const_cast<char*>(new_path.c_str()));  // NOLINT
-#endif
+    prepend_python_path(PYTHON_SITELIB);
 
     // On some python versions init_numpy() can throw an FPE, this occurred at
     // least with python 3.6, numpy 1.14.2.
@@ -80,6 +63,10 @@ SetupLocalPythonEnvironment::SetupLocalPythonEnvironment(
     disable_fpes.restore_exceptions();
   }
   initialized = true;
+
+  // Add test directory to the Python path
+  prepend_python_path(unit_test_src_path() +
+                      cur_dir_relative_to_unit_test_path);
 }
 
 #if PY_MAJOR_VERSION == 3

